pkg/sdn/plugin/node.go

```diff
@@ -125,18 +149,140 @@ func (node *OsdnNode) Start() error {
 			}
 		}
 	}
 
+	if err := node.cniServer.Start(); err != nil {
+		return err
+	}
+	go utilwait.Forever(node.watchCni, 0)
+
 	node.markPodNetworkReady()
 
 	return nil
 }
 
+type podWorker struct {
+	node     *OsdnNode
+	requests chan *podRequest
+}
+
+func newPodWorker(node *OsdnNode) *podWorker {
+	return &podWorker{
+		node: node,
+		// Allow queue depth of 10 before blocking additions, if there
+		// are more than 10 outstanding requests something is wrong
+		requests: make(chan *podRequest, 10),
+	}
+}
+
+func (w *podWorker) addRequest(request *podRequest) {
+	w.requests <- request
+	if len(w.requests) == 1 {
+		// Start proessing when the first request is added
+		go w.process()
+	}
+}
+
+func (w *podWorker) process() {
+	for request := range w.requests {
+		log.V(5).Infof("Handling pod network request %#v", request)
+
+		var err error
+		result := &podResult{code: http.StatusOK}
+		switch request.command {
+		case CNI_ADD:
+			var ipamResult *cnitypes.Result
+			ipamResult, err = w.node.setupPod(request)
+			if err == nil {
+				result.responseJson, err = json.Marshal(ipamResult)
+			}
+		case CNI_DEL:
+			err = w.node.teardownPod(request)
+		case CNI_UPDATE:
+			err = w.node.updatePod(request)
+		}
+
+		if err != nil {
+			result.code = http.StatusBadRequest
+			result.err = fmt.Sprintf("%v", err)
+		}
+		log.V(5).Infof("Returning pod network result %#v", result)
+		request.result <- result
+	}
+}
+
+func getPodKey(request *podRequest) string {
+	return fmt.Sprintf("%s/%s", request.podNamespace, request.podName)
+}
+
+func (node *OsdnNode) watchCni() {
+	// map of pod namespace+name to array of podWorkers
+	workers := make(map[string]*podWorker)
+
+	for {
+		select {
+		case request := <-node.podRequestChan:
+			log.V(5).Infof("Dispatching pod network request %#v", request)
+
+			podKey := getPodKey(request)
+			worker, ok := workers[podKey]
+			if !ok {
+				worker = newPodWorker(node)
+				workers[podKey] = worker
+			}
+
+			// Add the request to the worker
+			log.V(5).Infof("Added pod network request %#v", request)
+			worker.addRequest(request)
+
+			// Clean up finished workers
+			for key, worker := range workers {
+				// Safe to compare to 0 after as this is the only
+				// goroutine that adds things to the workers queue
+				if len(worker.requests) == 0 {
+					delete(workers, key)
+				}
+			}
+		}
+	}
+}
+
+// FIXME: this should eventually go into kubelet via a CNI UPDATE/CHANGE action
+// See https://github.com/containernetworking/cni/issues/89
+func (node *OsdnNode) UpdatePod(namespace string, name string, id kubeletTypes.ContainerID) error {
+	resultChan := make(chan *podResult)
+	req := &podRequest{
+		command:      CNI_UPDATE,
+		podNamespace: namespace,
+		podName:      name,
+		containerId:  id.String(),
+		result:       resultChan,
+	}
+
+	// Grab the netns from our cache since the Update request doesn't come from CNI
+	node.runningPodsLock.Lock()
+	runningPod, ok := node.runningPods[getPodKey(req)]
```

Comment:
Make runningPods its own type with safe accessors.  Multiple locks per type complicates reasoning about the effects of the lock and is generally better represented by encapsulating the locks in function calls. 

In general, if you have `.Unlock()` and it's not in a defer, you should rearrange your code until the lock and object under the lock until they are inside a function.